### ISSUE TITLE
Train image size 

### ISSUE THREAD

**Issue by osamaalschame** [ultralytics/ultralytics#5265]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How I can train with image size (height=1024,width=640).

**Comment by zshariff6506** [comment]
Interesting. I would look into Issue 3634 and https://stackoverflow.com/questions/75268393/yolov8-how-does-it-handle-different-image-sizes. Hope this helps!

**Comment by glenn-jocher** [comment]
@osamaalschame you can train your YOLOv8 model with a custom image size by specifying the image size in the model configuration file. In this file, there's a parameter called 'input_size' which is a list of two values, typically [height, width]. You should update these dimensions to your desired size, i.e. [1024, 640]. 

Keep in mind, however, that training with larger image sizes may require more computational resources. In case you face any issue with the memory limitation, you might consider using techniques like gradient accumulation to effectively train your model. 

Nevertheless, adjusting the input sizes to match your custom image size should allow you to train your YOLOv8 model effectively. Lastly, note that while you can train with these custom dimensions, the test-time inference might be more efficient with a square aspect ratio.

**Comment by haimat** [comment]
@glenn-jocher When I call `predict(img, imgsz=640)` with an `img` in the format width=3072 and height=2048, then YOLO prints this:

> 0: 448x640 (no detections), 59.4ms
> Speed: 4.8ms preprocess, 59.4ms inference, 0.7ms postprocess per image at shape (1, 3, 448, 640)

Does this mean that YOLO first prints the height of the image and then its width?

**Comment by glenn-jocher** [comment]
@haimat yes, you are correct. The first number (448 in your case) represents the height, and the second number (640 in your case) is the width. 

This means, when you see the numbers printed out as "448x640", YOLO is processing your image as 448 pixels in height and 640 pixels in width.

It's important to note that these dimensions may not be the same as the original image if you provided an 'imgsz' argument. If 'imgsz' is smaller than the original dimensions, YOLO will resize the image to match 'imgsz' while maintaining the original aspect ratio, padding the shorter dimension as required.

**Comment by shen865069799** [comment]
> @osamaalschame you can train your YOLOv8 model with a custom image size by specifying the image size in the model configuration file. In this file, there's a parameter called 'input_size' which is a list of two values, typically [height, width]. You should update these dimensions to your desired size, i.e. [1024, 640].
> 
> Keep in mind, however, that training with larger image sizes may require more computational resources. In case you face any issue with the memory limitation, you might consider using techniques like gradient accumulation to effectively train your model.
> 
> Nevertheless, adjusting the input sizes to match your custom image size should allow you to train your YOLOv8 model effectively. Lastly, note that while you can train with these custom dimensions, the test-time inference might be more efficient with a square aspect ratio.

@glenn-jocher  but in default.yaml<https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/default.yaml>, imgsz is [w,h]
![Screenshot from 2023-10-10 16-10-42](https://github.com/ultralytics/ultralytics/assets/31476124/f9d5ae75-3f2a-4529-98ec-49f75fd5a03c)
so, imgsz should be [h,w] ???

becase when I train with height(480) width(640), the parameter imgsz [640,480] 
and I export onnx with (model.export(format='onnx',imgsz=[640,480]),same to train),onnx input image shape is 1x3x640x480,but pytorch shape should [NxCxHXW]
![Screenshot from 2023-10-10 16-19-36](https://github.com/ultralytics/ultralytics/assets/31476124/c58e8d38-eb6e-4031-9c5e-3a5826feffc6)

**Comment by glenn-jocher** [comment]
@shen865069799, thanks for bringing this to our attention. If you observe a discrepancy in the results based on the input order, it might be due to a specific implementation detail of the exporting function in YOLOv8. Typically in YOLOv8, the order of the dimensions in the imgsz variable is [width, height], as you correctly noted in the default.yaml file.

However, when exporting the model, especially in formats like ONNX, the convention changes and it follows the order of [height, width].

To summarize, when you're training the model, use the order [width, height] for imgsz, but when exporting to ONNX, use the order [height, width] for imgsz.

Remember, when feeding the model with input images, the shape should maintain the format [NxCxHxW] as you noted.

I hope this helps clarify things. If you have more questions or if anything is unclear, feel free to ask. We appreciate your engagement with the project.

**Comment by jbdebelle** [comment]
> @osamaalschame you can train your YOLOv8 model with a custom image size by specifying the image size in the model configuration file. In this file, there's a parameter called 'input_size' which is a list of two values, typically [height, width]. You should update these dimensions to your desired size, i.e. [1024, 640].

Here you say that it should be [height, width] yet you follow by saying it should be [width, height] : 

> @shen865069799, thanks for bringing this to our attention. If you observe a discrepancy in the results based on the input order, it might be due to a specific implementation detail of the exporting function in YOLOv8. Typically in YOLOv8, the order of the dimensions in the imgsz variable is [width, height], as you correctly noted in the default.yaml file.

![image](https://github.com/ultralytics/ultralytics/assets/44498990/f72cfc14-c01d-4951-93b8-131a5a5be8d3)
and yet the documentation states that it is [height, width]
![image](https://github.com/ultralytics/ultralytics/assets/44498990/42a83b48-b628-41f4-b548-fbde52b882ec)

which one should we use ? 
Thanks

**Comment by glenn-jocher** [comment]
@jbdebelle apologies for the confusion earlier. The input size should be specified as [width, height] in the imgsz parameter during training. This is the convention followed by YOLOv8 in the model configuration file.

However, when exporting the model to formats like ONNX, the convention changes and you should switch the order to [height, width]. In this specific scenario, the format is different due to the export function's requirements.

So, in summary:
- For training, use [width, height]
- For exporting to ONNX, use [height, width]

Again, sorry for the confusion and I hope this clears things up. Let me know if you have any other questions or concerns!

**Comment by haimat** [comment]
@glenn-jocher Thanks for clarfying this.
Could you please mention that somewhere in your docs?

**Comment by glenn-jocher** [comment]
@haimat Thank you for your suggestion! We can see how specifying the [width, height] order for training and the [height, width] order for exporting to ONNX in the documentation would be helpful and prevent confusion among users of YOLOv8. We will take this into consideration and plan to update our documentation accordingly as soon as possible. We appreciate your feedback as it helps us improve the project. Please feel free to share any other suggestions or issues you might come across while using YOLOv8. Thank you once again for bringing this to our attention.

**Comment by haimat** [comment]
@glenn-jocher Great, thanks for your fast response!

**Comment by glenn-jocher** [comment]
@haimat you're welcome! I'm glad I could help. If you have any more questions or need further clarification, don't hesitate to ask. Happy coding!
